### DISCUSSION TITLE
Seeding with Pest without touching the main database

### DISCUSSION THREAD

**Discussion by spooksie** [pestphp/pest#427] [Q&A] [answered]
Hi,

I was previously using factories with random data to create my tests, however I hit a brick wall when trying to test with real data in the main database. 

At first I thought that Datasets was the cure, but after testing, it isn't what I needed. I dont need to test each row, I want to mimic my main database by adding real data before each test.

The problem is that the seed works, but adds it to the main database and not the test database (in memory).

**My Test**
```
<?php

namespace Tests\Feature\Livewire\Services;

use App\Models\am_product;
use function Pest\Livewire\livewire;
use App\Http\Livewire\Services\Products;
use Database\Seeders\am_product as amProducts;
use Illuminate\Foundation\Testing\RefreshDatabase;

uses(RefreshDatabase::class);

beforeEach(function () {
    $this->seed(amProducts::class);
});

it('can render the page', function () {
    $component = livewire(Products::class);
    $component->assertStatus(200);
});

it('check if it seeds', function () {
    ray(am_product::get());
});
```

**My phpunit.xml**
```
<?xml version="1.0" encoding="UTF-8"?>
<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:noNamespaceSchemaLocation="./vendor/phpunit/phpunit/phpunit.xsd"
         bootstrap="vendor/autoload.php"
         colors="true"
>
    <testsuites>
        <testsuite name="Unit">
            <directory suffix="Test.php">./tests/Unit</directory>
        </testsuite>
        <testsuite name="Feature">
            <directory suffix="Test.php">./tests/Feature</directory>
        </testsuite>
    </testsuites>
    <coverage processUncoveredFiles="true">
        <include>
            <directory suffix=".php">./app</directory>
        </include>
    </coverage>
    <php>
        <server name="APP_ENV" value="testing"/>
        <server name="BCRYPT_ROUNDS" value="4"/>
        <server name="CACHE_DRIVER" value="redis"/>
        <server name="DB_CONNECTION" value="sqlite_testing"/>
        <server name="DB_DATABASE" value=":memory:"/>
        <server name="MAIL_MAILER" value="array"/>
        <server name="QUEUE_CONNECTION" value="sync"/>
        <server name="SESSION_DRIVER" value="array"/>
        <server name="TELESCOPE_ENABLED" value="false"/>
    </php>
</phpunit>
```

**Comment by nunomaduro** [comment]
cc @lukeraymonddowning Do you see the reason? 🤔

**Reply by spooksie** [reply]
Ive had a hard day wrestling with the test file. 

I found a way to seed that did what I needed it to do with 

```
beforeEach(function () {
    $this->seed(services_categories::class);
    $this->seed(services_panel::class);
    $this->seed(services_products::class);
    $this->seed(services_relations::class);
    $this->seed(services_sites::class);
    $this->seed(panels::class);
    $this->seed(panel_services::class);
});
```

Worked an passed the tests, but....
**uses(RefreshDatabase::class);**
Decided to delete the main database and not the one in memory. Luckily I had a backup.

I've removed the db_connection and db_database lines in phpunit.xml and created a .env.testing file which I changed the db name to include _testing at the end. Ironclad was the main db name.

```
...
DB_CONNECTION=mysql
DB_HOST=127.0.0.1
DB_PORT=3306
DB_DATABASE=ironclad_testing
DB_USERNAME=root
DB_PASSWORD=
...
```

I created the new ironclad_testing inside phpmyadmin. Its doing the same and wiping the main db and ignoring ironclad_testing

I've tried previous tests and its doing the same. Plus I updated PestPhp this morning. 

Could this be a bug or am I missing something?

**Reply by lukeraymonddowning** [reply]
This is strange. I'm not able to replicate locally 🤔

Sounds like it might be something deeper, but no clue what

**Reply by spooksie** [reply]
Its weird, as its ignoring .env.testing and also phpunit.xml variables. Its like its running on production. 

Im going to do some more digging today.

**Reply by nunomaduro** [reply]
Thank you @spooksie

**Comment by spooksie** [comment, marked as answer]
I added this to my test
`ray(env('APP_ENV'), env('DB_CONNECTION'), env('DB_DATABASE'));`

It was showing my main connection, so this makes sense. After googling this there is an artisan command called config:clear. I ran the command and tried again. It was showing my main connection again.

How I fixed it in the end
1. php artisan config:clear
2. Restart Valet
3. Restart Mac
4. php artisan config:clear again
5. Ran the test again and it worked with the correct database. 

Issue solved for now, looks like the caching was causing issues. I don't know if its a Laravel issue or Valet issue.
